app/services/SubmitMessageService.scala

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+import java.time.OffsetDateTime
+
+import connectors.MessageConnector
+import javax.inject.Inject
+import play.api.Logger
+import models.Arrival
+import models.ArrivalId
+import models.ArrivalState
+import models.MessageId
+import models.MessageState
+import models.MovementMessageWithState
+import models.SubmissionResult
+import repositories.ArrivalMovementRepository
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+
+class SubmitMessageService @Inject()(
+  arrivalMovementRepository: ArrivalMovementRepository,
+  messageConnector: MessageConnector
+)(implicit ec: ExecutionContext) {
+
+  private val logger = Logger(getClass)
+
+  def submit(arrivalId: ArrivalId, messageId: MessageId, message: MovementMessageWithState)(implicit hc: HeaderCarrier): Future[SubmissionResult] =
+    arrivalMovementRepository.addNewMessage(arrivalId, message) flatMap {
+      case Failure(_) =>
+        Future.successful(SubmissionResult.FailureInternal)
+
+      case Success(_) => {
+        messageConnector
+          .post(arrivalId, message, OffsetDateTime.now)
+          .flatMap {
+            _ =>
+              arrivalMovementRepository
+                .setArrivalStateAndMessageState(arrivalId, messageId, ArrivalState.ArrivalSubmitted, MessageState.SubmissionSucceeded)
+                .map {
+                  _ =>
+                    SubmissionResult.Success
+                }
+                .recover({
+                  case _ =>
+                    SubmissionResult.FailureInternal
+                })
+          }
+          .recoverWith {
+            case error => {
+              logger.error(s"Existing Movement - Call to EIS failed with the following Exception: ${error.getMessage}")
```

Comment:
I think for both of these comments we may want to look at making the API of the repository consistent since adding a new movement returns a `Future[Unit]` and the adding a new message returns a `Future[Try[Unit]]`. So in the former, we've decided to fail the future and forced us to use recover blocks whereas in the other we return a `Try` where the exception is one that we throw. I think unifying the those two APIs, in particular, will let us see even more commonality between these two methods and allow us to extract even more. I'm also interested in removing the throw Exception behaviour from the connector level and getting rid of the recover blocks that deal with those, so that may affect these also.

Do ou have any thoughts on the impact of these changes? I think if they're made there will be a need to refactor this again so the code can change, so I'm hesitant to refactor now

Reply:
No problem, makes sense (think I overlooked the different return types) Happy for you to hold off on this no problem